embed/extmod/modtrezorcrypto/modtrezorcrypto-secp256k1_zkp.h

```diff
@@ -0,0 +1,247 @@
+
+/// def verify(public_key: bytes, signature: bytes, digest: bytes) -> bool:
+///     '''
+///     Uses public key to verify the signature of the digest.
+///     Returns True on success.
+///     '''
+STATIC mp_obj_t mod_trezorcrypto_secp256k1_zkp_verify(mp_obj_t public_key, mp_obj_t signature, mp_obj_t digest) {
+    const secp256k1_context *ctx = mod_trezorcrypto_secp256k1_context();
+    mp_buffer_info_t pk, sig, dig;
+    mp_get_buffer_raise(public_key, &pk, MP_BUFFER_READ);
+    mp_get_buffer_raise(signature, &sig, MP_BUFFER_READ);
+    mp_get_buffer_raise(digest, &dig, MP_BUFFER_READ);
+    if (pk.len != 33 && pk.len != 65) {
+        mp_raise_ValueError("Invalid length of public key");
+    }
+    if (sig.len != 64 && sig.len != 65) {
+        mp_raise_ValueError("Invalid length of signature");
+    }
+    int offset = sig.len - 64;
+    if (dig.len != 32) {
+        mp_raise_ValueError("Invalid length of digest");
+    }
+    secp256k1_ecdsa_signature ec_sig;
+    if (!secp256k1_ecdsa_signature_parse_compact(ctx, &ec_sig, (const uint8_t *)sig.buf + offset)) {
+        mp_raise_ValueError("Invalid signature");
+    }
+    secp256k1_pubkey ec_pk;
+    if (!secp256k1_ec_pubkey_parse(ctx, &ec_pk, (const uint8_t *)pk.buf, pk.len)) {
+        mp_raise_ValueError("Invalid public key");
+    }
+    return mp_obj_new_bool(1 == secp256k1_ecdsa_verify(ctx, &ec_sig, (const uint8_t *)dig.buf, &ec_pk));
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(mod_trezorcrypto_secp256k1_zkp_verify_obj, mod_trezorcrypto_secp256k1_zkp_verify);
+
+/// def verify_recover(signature: bytes, digest: bytes) -> bytes:
+///     '''
+///     Uses signature of the digest to verify the digest and recover the public key.
+///     Returns public key on success, None on failure.
+///     '''
+STATIC mp_obj_t mod_trezorcrypto_secp256k1_zkp_verify_recover(mp_obj_t signature, mp_obj_t digest) {
+    const secp256k1_context *ctx = mod_trezorcrypto_secp256k1_context();
+    mp_buffer_info_t sig, dig;
+    mp_get_buffer_raise(signature, &sig, MP_BUFFER_READ);
+    mp_get_buffer_raise(digest, &dig, MP_BUFFER_READ);
+    if (sig.len != 65) {
+        mp_raise_ValueError("Invalid length of signature");
+    }
+    if (dig.len != 32) {
+        mp_raise_ValueError("Invalid length of digest");
+    }
+    int recid = ((const uint8_t *)sig.buf)[0] - 27;
+    if (recid >= 8) {
+        mp_raise_ValueError("Invalid recid in signature");
+    }
+    bool compressed = (recid >= 4);
+    recid &= 3;
+
+    secp256k1_ecdsa_recoverable_signature ec_sig;
+    if (!secp256k1_ecdsa_recoverable_signature_parse_compact(ctx, &ec_sig, (const uint8_t *)sig.buf + 1, recid)) {
+        mp_raise_ValueError("Invalid signature");
+    }
+    secp256k1_pubkey pk;
+    if (!secp256k1_ecdsa_recover(ctx, &pk, &ec_sig, (const uint8_t *)dig.buf)) {
+        // Copying "return None" behaviour from secp256k1 module; but why do
+        // we do this rather than raising an exception?
```

Comment:
Looks like a TODO.

Reply:
Yes, this is still open. We raise an exception for every other failure mode, so I see no reason why we should not raise an exception here. Any ideas?

Reply:
Well, so the documentation of the function says that it should return `None` if the signature is invalid, so `None` makes sense. However, the actual behavior of the function is that either returns `None` or raises `ValueError` if the signature is invalid.

This inconsistency lead to a bug (#422), which was  fixed in https://github.com/trezor/trezor-core/commit/d5fb2a477a7267ec6c88b7e09c39c8c9a63022fc by handling both `None` and an exceptions equally in the caller code. (The Ethereum app is the only caller.)

I think this can be improved by making the function consistently return `None` no matter what the failure is, but in a different PR. I will remove your comment then at least here. If you agree with this conclusion, I'm volunteering to create that other PR, it should just take a few minutes.